### ISSUE TITLE
Does not work on micro:bit V2 hardware

### ISSUE THREAD

**Issue by timleeanieducation** [bsiever/microbit-dstemp#2]
The dstemp extension does not work on the micro:bit V2. It returns an error code 3 "Read Timeout".

**Comment by bsiever** [comment]
@timleeanieducation  Just for more info, does it work on v1 for you?    

Also, where did you get your sensor?  There are a variety of copies that have different quirks.  So far I've only tested it on sensors that are from a reliable supplier.

I my tests with my sensors, it worked well on both V1 and V2 hardware.

**Comment by timleeanieducation** [comment]
It works perfectly fine for V1, using the latest stable micro:bit makecode, with the V2 default pull fix (https://github.com/lancaster-university/codal-microbit-v2/issues/17).

You may be right about the sensor supplier. I'll get some DS18B20 for testing from DigiKey, which are manufactured by Maxim Integrated.

I'm also wondering if the V2's sink current capability may have something to do with it. V2 pins (P0 to P2) are not capable of sinking 4mA.

**Comment by bsiever** [comment]
Thanks for the update and the additional testing!  I have tested it on a v2 and it worked...but I'll do more testing as time permits.  

I have an alternate version for 2-wire mode, which I've also tested on the v2...but if this (3-wire) doesn't work, I'm pretty skeptical that it will.  None the less, it's listed as microbit-dstemp-2wire.  The repo shows the wiring for 2-pin mode: https://github.com/bsiever/microbit-dstemp-2wire

**Comment by bsiever** [comment]
@timleeanieducation  I just did some additional testing:  I had two sensors (one in 3-wire mode and one in 2-wire mode) and two V2 micro:bits.  Both micro:bits worked with each sensor and the respective extensions.  The 3-wire setup has a 1M cable and uses a breadboard.  The 2-wire version ist just the sensor and alligator clips. 

Are you able to share where you got your sensor?  I may be able to order one and try it out.   

My sensors came from http://yourduino.com/.  At the time I ordered them (~1.5 years ago) they had a reputation for supplying authentic DS18B20s.   Here's a list of reliable sources: https://github.com/cpetrich/counterfeit_DS18B20#sources

**Comment by timleeanieducation** [comment]
The cpetrich link is very useful. I downloaded the code to test my sensor. This is what I got:

1. discover_fake_DS18B20.ino
28-CC-3D-75-D0-01-3C-7D: ROM does not follow expected pattern 28-xx-xx-xx-xx-00-00-crc. Error.
Scratchpad Register: 90/01/55/05/7F/A5/A5/66/B4
Info only: Scratchpad bytes 2,3,4 (55/05/7F):  not Maxim default values 4B/46/7F.
Scratchpad byte 5 (0xA5):  should have been 0xFF according to datasheet. Error.
Scratchpad byte 6 (0xA5):  unexpected value. Error.
Scratchpad byte 7 (0x66):  should have been 0x10 according to datasheet. Error.
0x4E modifies alarm registers:  ok.
0x4E accepts 10 bit resolution:  ok.
0x4E preserves reserved bytes:  ok.
0x4E accepts 12 bit resolution:  ok.
0x4E preserves reserved bytes:  ok.
Checking byte 6 upon temperature change: not necessary. Skipped.
--> Sensor appears to be counterfeit based on 4 deviations.

2. classify_fake_DS18B20.ino
28-CC-3D-75-D0-01-3C-7D: Family D2 (Clone w/o parasitic power).

The sensor is encpasulated in waterproof shield, purchased from TaoBao, and obviously fake. There are hundreds of shops selling them, and they are really cheap (around 3 to 5 RMB). What I find intriguing is why it worked perfectly on micro:bit V1 but not on V2.

I have ordered a 3-wire and a 2-wire DS18B20 from DigiKey to test things out. Will keep you informed of progress.

**Comment by bsiever** [comment]
@timleeanieducation Thanks for the extra testing!    The sensors require pretty precise timing, which I had to "hand tune" based on the processor speed. So the code used for the v2 is a little different than the v1.  It's possible my "tuning" is just a little off for your sensor. 

I'll try to do a few tests with fake sensors when time permits.  Hopefully I can update the extensions so it'll typically work for nearly everyone.  (That being said, I think the parasitic power / 2-wire mode will be great for many people and the fake sensors don't seem to support it very well)

**Comment by timleeanieducation** [comment]
A DS18B20 and DS18B20-PAR arrived today from DigiKey. Tested using discover_fake_DS18B20.ino and micro:bit V1/V2.
bsiever/microbit-dstemp 0.1.21
bsiever/microbit-dstemp-2wire 0.1.6

DS18B20:
28-57-9F-F1-0C-00-00-28: ROM ok.
  Scratchpad Register: 88/01/4B/46/7F/FF/08/10/76
  Info only: Scratchpad bytes 2,3,4 (4B/46/7F):  Maxim default values.
  Scratchpad byte 5 (0xFF):  ok.
  Scratchpad byte 6 (0x08):  ok.
  Scratchpad byte 7 (0x10):  ok.
  0x4E modifies alarm registers:  ok.
  0x4E accepts 10 bit resolution:  ok.
  0x4E preserves reserved bytes:  ok.
  0x4E accepts 12 bit resolution:  ok.
  0x4E preserves reserved bytes:  ok.
  Checking byte 6 upon temperature change: not necessary. Skipped.
  --> Sensor responded like a genuie Maxim.
Both micro:bit V1 and V2 produced correct temperature.

DS18B20-PAR:
28-E8-CC-FD-0C-00-00-C2: ROM ok.
  Scratchpad Register: FF/07/4B/46/7F/FF/01/10/2F
  Info only: Scratchpad bytes 2,3,4 (4B/46/7F):  Maxim default values.
  Scratchpad byte 5 (0xFF):  ok.
  Scratchpad byte 6 (0x01):  ok.
  Scratchpad byte 7 (0x10):  ok.
  0x4E modifies alarm registers:  ok.
  0x4E accepts 10 bit resolution:  ok.
  0x4E preserves reserved bytes:  ok.
  0x4E accepts 12 bit resolution:  ok.
  0x4E preserves reserved bytes:  ok.
  Checking byte 6 upon temperature change:  Inconclusive. Please change sensor temperature and repeat.
  --> DS18S20 counterfeit test not completed, otherwise sensor responded like a genuie Maxim.
micro:bit V1 produced correct temperature. V2 output is -Infinity.

**Comment by bsiever** [comment]
@timleeanieducation  Wow!  Thanks for all the testing.  It looks like I definitely need to do more testing review with the DS18B20-PAR on the v2....but I still plan to do more testing with "fakes" too.   Just to check:  did you test either/both sensors with just 3-wire mode (the plain "dstemp" extension) or did you test either/both with the 2-wire mode too (dstemp-2wire)?

Thanks again --- this really helps!

**Comment by bsiever** [comment]
@timleeanieducation Ah, an update.   I didn't realize what the difference was between the PAR and non-PAR versions.  The non-PAR version should be able to do either parasitic power or non-parasitic power.  Hopefully it will work with either the dstemp or dstemp-2wire extensions if wired correctly. 

The PAR version is already internally wired in parasitic power mode, so it would only be expected to work with the dstemp-2wire extension.  (See: https://www.maximintegrated.com/en/design/technical-documents/app-notes/2/203.html).

This doesn't explain why the PAR would work on v1 and not v2 --- it may be the difference in the timing (which I can change), or power, or the internal resistance of the v2 (the internal resistance on both the v1 and v2 is technically out-of-spec for the DS18B20).

**Comment by timleeanieducation** [comment]
Thanks @bsiever for the detailed information. I did not realize you can do a 2-wire conection on a regular DS18B20. I have re-tested all my devices making sure they are wired correctly and using the correct extensions. The results are shown below:

![ds18b20_test_results_2021-05-05](https://user-images.githubusercontent.com/17022515/117104701-34b39b00-adaf-11eb-9ba8-773318dab9b9.png)

Wiring diagrams
DS18B20 External Powered Circuit
![DS18B20-External-Powered-Circuit](https://user-images.githubusercontent.com/17022515/117098399-f57e4d80-ada0-11eb-96af-fd1ef31b3e35.png)

DS18B20 Parasite Powered Circuit
![DS18B20-Parasite-Powered-Circuit](https://user-images.githubusercontent.com/17022515/117099529-ff558000-ada3-11eb-9a2a-d935675777d9.png)

DS18B20-PAR Parasite Powered Circuit
![DS18B20-PAR-Parasite-Powered-Circuit](https://user-images.githubusercontent.com/17022515/117099545-08465180-ada4-11eb-965d-f6fefcf5b5a6.png)

**Comment by timleeanieducation** [comment]
Just a side note, the batch of sensors I bought from the same shop in TaoBao appear to be three different versions. Under the 3-wire mode and using bsiever/microbit-dstemp 0.1.21 they work for micro:bit V1 but produced a "Read Timeout" error for micro:bit V2. They might be the same chip but with different ROM codes.

TaoBao Unshielded
28-E7-4D-75-D0-01-3C-5E: ROM does not follow expected pattern 28-xx-xx-xx-xx-00-00-crc. Error.
Scratchpad Register: 90/01/55/05/7F/A5/81/66/4E
Info only: Scratchpad bytes 2,3,4 (55/05/7F):  not Maxim default values 4B/46/7F.
Scratchpad byte 5 (0xA5):  should have been 0xFF according to datasheet. Error.
Scratchpad byte 6 (0x81):  unexpected value. Error.
Scratchpad byte 7 (0x66):  should have been 0x10 according to datasheet. Error.
0x4E modifies alarm registers:  ok.
0x4E accepts 10 bit resolution:  ok.
0x4E preserves reserved bytes:  ok.
0x4E accepts 12 bit resolution:  ok.
0x4E preserves reserved bytes:  ok.
Checking byte 6 upon temperature change: not necessary. Skipped.
--> Sensor appears to be counterfeit based on 4 deviations.

TaoBao Shielded
28-B1-98-07-D6-01-3C-30: ROM does not follow expected pattern 28-xx-xx-xx-xx-00-00-crc. Error.
Scratchpad Register: 80/01/55/05/7F/A5/81/66/1A
Info only: Scratchpad bytes 2,3,4 (55/05/7F):  not Maxim default values 4B/46/7F.
Scratchpad byte 5 (0xA5):  should have been 0xFF according to datasheet. Error.
Scratchpad byte 6 (0x81):  unexpected value. Error.
Scratchpad byte 7 (0x66):  should have been 0x10 according to datasheet. Error.
0x4E modifies alarm registers:  ok.
0x4E accepts 10 bit resolution:  ok.
0x4E preserves reserved bytes:  ok.
0x4E accepts 12 bit resolution:  ok.
0x4E preserves reserved bytes:  ok.
Checking byte 6 upon temperature change: not necessary. Skipped.
--> Sensor appears to be counterfeit based on 4 deviations.

TaoBao with Pluggable Terminal
28-CC-3D-75-D0-01-3C-7D: ROM does not follow expected pattern 28-xx-xx-xx-xx-00-00-crc. Error.
Scratchpad Register: 90/01/55/05/7F/A5/A5/66/B4
Info only: Scratchpad bytes 2,3,4 (55/05/7F): not Maxim default values 4B/46/7F.
Scratchpad byte 5 (0xA5): should have been 0xFF according to datasheet. Error.
Scratchpad byte 6 (0xA5): unexpected value. Error.
Scratchpad byte 7 (0x66): should have been 0x10 according to datasheet. Error.
0x4E modifies alarm registers: ok.
0x4E accepts 10 bit resolution: ok.
0x4E preserves reserved bytes: ok.
0x4E accepts 12 bit resolution: ok.
0x4E preserves reserved bytes: ok.
Checking byte 6 upon temperature change: not necessary. Skipped.
--> Sensor appears to be counterfeit based on 4 deviations.

**Comment by bsiever** [comment]
@timleeanieducation  Thanks again...I think I spotted a difference between your testing and mine. Did you use an external pullup resistor (the 4.7k) for the 2wire?  It shouldn't need it. I enable the _internal_ pull-up resistors on the micro:bit's processor.  All of my testing of 2wire omitted the external resistor.

The 2-wire repo has an example of how I wired temperature probes: https://github.com/bsiever/microbit-dstemp-2wire#building-a-sensor-assembly  (It's literally two wires and the sensor probe. No other parts at all.  That was my actual goal: to minimize parts/cost and make it easier to build a micro:bit compatible probe)

If you get a chance, it'd be interesting to see if removing the resistor has an impact (it would change the net resistance and have some impact on total power and the effort to "pull")

**Comment by timleeanieducation** [comment]
Interesting... I have that in mind too. Likewise, I would like to minimize parts to save cost. Since I'm going to test for internal pull up, I might as well do the tests where there's no external resistor. Here are the results:

![ds18b20_test_results_2021-05-06](https://user-images.githubusercontent.com/17022515/117253227-92acb500-ae79-11eb-8d01-104cd6ad5a97.png)

Conclusions:
1. It seems for the DS18B20-PAR sensor, it will work with external, no external and internal pull up.
2. The same is true for DS18B20 in 2-wire configuration.
3. For 3-wire configuration, an external pull up resistor is required.
4. The surprising result is when 3-wire is used with internal pull up, which will work in V2 but not in V1.

**Comment by timleeanieducation** [comment]
Update: I bought another batch of waterproof DS18B20 sensors from TaoBao because it came with better wiring.

![Qhebot DS18B20 Sensor](https://user-images.githubusercontent.com/17022515/117749566-9313ca00-b244-11eb-8f3d-d7efec8e4285.png)

The ROM code is different yet again:

TaoBao with Qhebot Pluggable Terminal
28-60-30-AF-5D-20-01-9E: ROM does not follow expected pattern 28-xx-xx-xx-xx-00-00-crc. Error.
  Scratchpad Register: 50/05/4B/46/7F/FF/7F/10/FF
  Info only: Scratchpad bytes 2,3,4 (4B/46/7F):  Maxim default values.
  Scratchpad byte 5 (0xFF):  ok.
  Scratchpad byte 6 (0x7F):  unexpected value. Error.
  Scratchpad byte 7 (0x10):  ok.
  0x4E modifies alarm registers:  ok.
  0x4E accepts 10 bit resolution:  ok.
  0x4E preserves reserved bytes:  no, got: FF/3F/10. Error.
  0x4E accepts 12 bit resolution:  ok.
  0x4E preserves reserved bytes:  no, got: FF/7F/10. Error.
  Checking byte 6 upon temperature change: not necessary. Skipped.
  --> Sensor appears to be counterfeit based on 4 deviations.

This time, it works for both micro:bit V1 and V2 using the 3-wire bsiever/microbit-dstemp 0.1.21 extension.

**Comment by bsiever** [comment]
@timleeanieducation  Nice module --- it looks like it has the pull-up resistor built in!   I wanted to be able to connect the sensor with alligator (or crocodile) clips.  Combining that module with these clips (Source of image below and clips: [https://www.amazon.com/gp/product/B06XX2JBNC](https://www.amazon.com/gp/product/B06XX2JBNC)) would be an easy approach:
![image](https://user-images.githubusercontent.com/1421446/117811667-723c7c00-b226-11eb-8c01-a58b7bfdbaa8.png)

**Comment by timleeanieducation** [comment]
I have that as well. Together they cost around 25 RMB per set (3.89 USD, 30.2 HKD) which is a good price for student work. The pluggable module has a 4.7k pull up resistor and a 4.7k + LED indicator across Vcc and GND.  It appears Qhebot is their own brand. You can find the product here:

https://item.taobao.com/item.htm?spm=a230r.1.14.102.3d771f5bBuhCM8&id=637832887998&ns=1&abbucket=5#detail

The TaoBao website is for PRC domestic consumers, please use Google Translate.

**Comment by dongjoon9331** [comment]
The error in v2 seems not solved yet.
I run a program to educate teachers. I bought a cheap version of DS18B20 and am using it. It seems that the error issue has not been resolved in v2, so I am asking for a favor.

https://javalab.org/microbit_ds18b20/

**Comment by bsiever** [comment]
@dongjoon9331 At this time I only feel able to support official sensors from Dallas Semiconductor / Maxim.  There are a lot of "fake" sensors sold as DS18B20 that behave in unusual ways and I don't have examples of them to test or the time to ensure they'll all work.  If possible, please only use sensors from known sources.  

You'll notice in the messages above there are some tips to test sensors using and Arduino and @timleeanieducation has done extensive testing on several sensors.   

Also, if you have access to official DS18B20 sensors, you may want to try the other version of this extension at: https://github.com/bsiever/microbit-dstemp-2wire  .  It uses the "parasitic power mode" of the sensor and can be connected with just two wires (and no external resistor).  It may be an easier option for some users.  The repository's [README.md](https://github.com/bsiever/microbit-dstemp-2wire/blob/master/README.md) includes directions to assemble a sensor.

**Comment by dongjoon9331** [comment]
@bsiever My sensor is also fake. Since our country is close to China, similar phenomenons are found. Of course, there is also the impact of low prices.
Nevertheless, we do not require high precision and reliability that came from genuine parts. I just want to use it for secondary school education. Wouldn't it be possible to support legacy parts by removing a few error detection codes? The error detection code seems to be written in 'dstemp.cpp'.


![Image1](https://user-images.githubusercontent.com/75997501/126423999-ec12e200-ef4f-4d13-b538-0b17a5f8706d.png)


![Image2](https://user-images.githubusercontent.com/75997501/126424006-2c164a2f-bad5-4630-84b5-a2edc72b856d.png)


![Image3](https://user-images.githubusercontent.com/75997501/126424013-dfe070ef-3a4c-4c7c-a068-8b0f161aedc0.png)

**Comment by bsiever** [comment]
@dongjoon9331 It is not that simple.   I don't think there's a "half way" that's meaningful.  It already attempts to read the temperature three times just in case there's a small, rare error.  Without access to your specific sensor I have no way of knowing how values would need to be changed to cause it to work. For example, I don't know if the timing is too fast or too slow.  

I originally developed this extension for use in science school experiments and I used the cheapest sensors I could find, which were also fake.  They often different by ~2 degrees Celsius from one sensor to another, which was enough to be a problem for many experiments. Genuine sensors seem to vary by about 0.1 degree.  I feel genuine sensors are worth the price.  For me they were ~$4 U.S. rather than $2 for fakes.  Although doubling the price, this is still significantly less than the micro:bit itself or most other micro:bit accessories. 

I'll try to ensure that this extension continues to work with the DS18B20 sensors that it's designed for --- I've selected things based on the official data sheet for the DS18B20.

**Comment by dongjoon9331** [comment]
@bsiever  
Thank you. I gonna try with more sensors. 
Anyway, I updated my blog for Korean science teachers. https://javalab.org/microbit_ds18b20/

**Comment by dongjoon9331** [comment]
@bsiever 
The source shows that "readScratchpad(gpio, temp)" returns false for the fake DS18B20.
It seems that the  "readScratchpad(gpio, temp)" function returns a true value whenever pass the three steps (CRC, HIGH_ALARM, LOW_ALARM). 
I think it's possible with a little looser setting for this part.
I'd appreciate it if you could test that part.

![Image2](https://user-images.githubusercontent.com/75997501/126580556-c6ef636a-451c-4575-a2d2-438113030b14.png)

![Image3](https://user-images.githubusercontent.com/75997501/126580566-f57aed4a-007b-4d3f-b34b-a2e95d083897.png)

**Comment by bsiever** [comment]
@dongjoon9331 There are multiple parameters that could be critical here (especially `TIME_RECOV` and `TIME_SLOT`).  Without a sensor to test on, I don't know which or how to change them.  You are welcome to fork this repo and try changes.  If you find a fix and it doesn't adversely impact actual DS18B20 behavior, I'm open to a Pull Request and acknowledging your contributions.

As I said before, I will try my best to support authentic DS18B20.  I don't have the resources to support other variants (the ones I've tested didn't perform well anyway).

**Comment by dongjoon9331** [comment]
@bsiever 
I bought a sensor from DIGI-KEY. 
Genuine test connected to Arduino has passed. 
However, the v2 board still transmits an error message. 
The difference from the past is that the correct value transmits intermittently.
And same error in parasite mode too.

![20210728_171308](https://user-images.githubusercontent.com/75997501/127289865-5c1b0723-5610-4162-b87a-a5d2079caa77.jpg)
![Image1](https://user-images.githubusercontent.com/75997501/127289899-42e768b6-717c-4701-b812-2335527548cf.png)

![20210728_171700](https://user-images.githubusercontent.com/75997501/127289972-2911d653-73dd-450c-90be-6e637848d11f.jpg)
![Image2](https://user-images.githubusercontent.com/75997501/127289989-68f8fcb0-c02d-49c8-8731-29aa28113a67.png)
![Image3](https://user-images.githubusercontent.com/75997501/127290002-fc3a81f1-ff45-412d-91d5-f92257f96d4c.png)

**Comment by bsiever** [comment]
@dongjoon9331  I'll try my best to help with this problem:

1. Can you confirm that the resistor is connected solidly?  (That is, that it's either soldered or that wire is twisted around it and also firmly connected to the micro:bit).  
2. Can you test this project, which includes an additional error message: https://makecode.microbit.org/_4JiR5YfTd3g9
2. Would you be able to also test the other extension I've written, which doesn't require the resistor?  It may provide some additional insight and it's usually easier to use. 
    1. Here's the wiring: https://github.com/bsiever/microbit-dstemp-2wire#building-a-sensor-assembly (Twist red and black together and connect to GND; Connect white to P0).    
    2. Here's a complete project you can use: https://makecode.microbit.org/_9JJiPx7JPJ2r

**Comment by dongjoon9331** [comment]
@bsiever 
1. The resistor is fully connected with the Molex-5051 connector using a hand crimping tool.

2. Some test results are blows. (only v1.5 board with 3-wire connection worked correctly)

![Image5](https://user-images.githubusercontent.com/75997501/127334950-caeedf68-f788-467f-8db9-238aa084c98f.png)

![Image6](https://user-images.githubusercontent.com/75997501/127334965-da4cb12a-56f2-4530-9d8d-bc345f6c3fea.png)

![Image7](https://user-images.githubusercontent.com/75997501/127335760-58b3e1be-bb9a-4399-b774-7142fc24fe6b.png)

![Image8](https://user-images.githubusercontent.com/75997501/127334990-6dd041a6-8ae1-4fc4-8fd7-713a8f2b9d7b.png)

**Comment by bsiever** [comment]
@dongjoon9331 Your sensors are failing with "Not Connected", which means the sensor isn't replying correctly when it is reset.  I do not experience this with my sensors, so I have no easy way to be sure of the problem.  I need you to provide a bit more info.  

It looks like you've already uncommented line 25 of dstemp.cpp (so it now says `#define DEBUG 1`).  Can you re-run your experiments using your version of the repo (add the URL rather than searching for mine).  You'll get some additional messages about `pres=???` and `release=???`.  I suspect that `pres` will be 1 but `release` will be 0.  If so, I think this first error is a timing issue that we can probably compensate for.   That may only be the first problem though.   Please let me know what you find.

**Comment by bsiever** [comment]
@timleeanieducation Hi --- I finally did some testing with a "fake" DS18B20 and adjusted the timing a bit to work with my one fake sensor.  It may work with some of your collection of fakes now too.

**Comment by dongjoon9331** [comment]
@bsiever 
Hi! Now it works fine. thank you. However, some 'fake' sensors are returning '-inf'. There seems to be no problem if you add and use the error detection code as shown below.
[https://makecode.microbit.org/81264-91201-89352-30243](https://makecode.microbit.org/81264-91201-89352-30243)

**Comment by bsiever** [comment]
@timleeanieducation Thanks for checking!   I want to make sure I understand...is this correct: You've got some sensors that often return the correct temperature, but sometimes they return -Inf.  ?

**Comment by dongjoon9331** [comment]
@bsiever check it out, please.
![Image2](https://user-images.githubusercontent.com/75997501/140253487-8a4969e5-013f-45ca-8b56-c485cf5b7f95.png)
![Image3](https://user-images.githubusercontent.com/75997501/140253494-188638a8-ae34-4ffb-af48-abe725872cd1.png)

**Comment by timleeanieducation** [comment]
Hi @bsiever, I have finished testing my collection of DS18B20 sensors. But before going into the results, just a recap of my previous tests. I have a genuine DS18B20 sensor purchased from DigiKey, a few from Qhebot and quite a number with various implementations from TaoBao.

Using microbit-dstemp 0.1.21, the genuine DS18B20 and Qhebot worked for both micro:bit V1 and V2, as mentioned in my May 6th and 11th tests.

This round of tests uses microbit-dstemp 0.1.22 and the following sensors (details in the attachment).
[DS18B20 Devices.txt](https://github.com/bsiever/microbit-dstemp/files/7479973/DS18B20.Devices.txt)
Device ROM: 
DS18B20:  28-57-9F-F1-0C-00-00-28
Qhebot1:  28-FA-51-E2-5D-20-01-41
Qhebot2:  28-8D-A5-0C-5E-20-01-5D
Shielded1:  28-B1-98-07-D6-01-3C-30
Shielded2:  28-0F-48-07-D6-01-3C-7A
Unshielded1:  28-E7-4D-75-D0-01-3C-5E
Unshielded2:  28-15-8A-75-D0-01-3C-31
Module1:  28-3F-1E-75-D0-01-3C-E0
Module2:  28-14-40-75-D0-01-3C-99

![ds18b20_test_results_2021-11-05](https://user-images.githubusercontent.com/17022515/140445858-3fde8cc5-f037-4559-91e8-0ef5d3fad0e0.png)

So in other words, the genuine and Qhebot sensors for V2 worked using microbit-dstemp 0.1.21, but failed using microbit-dstemp 0.1.22. The surprise was the other TaoBao sensors for V2 failed using microbit-dstemp 0.1.21, but worked using microbit-dstemp 0.1.22.

**Comment by bsiever** [comment]
@timleeanieducation  It's frustrating that it worked with many "fakes", but not your genuine sensors.  I've updated it yet again in the hopes of making it work with everything.  

I sincerely appreciate all the testing you've done!  If you have extra of each type of sensor, I'm willing to pay you for the sensors and costs to ship them to me.  That would allow me to do this testing (and, more importantly, see the details when things fail).

Thanks again...I hope this patch restores the genuine sensors to "working".

**Comment by timleeanieducation** [comment]
Hi @bsiever, sorry for not replying sooner. I do have some spare sensors that I can send to you for testing. Please send your postal address to tim.lee@anieducation.com. Don't worry about the postage. We shall pay for it, it's the least we could do.

**Comment by CAybi777** [comment]
![Image](https://github.com/user-attachments/assets/06de16d7-1325-4f64-9bbc-ad3b102bf5d6)
Presento el mismo error, las mediciones me salen '-infinity' la única solución que conseguí fue el termómetro inalambrico de javalab del coreano. Quisiera saber si puede habilitar la opción de trabajar almenos dos cifras decimales en lugar de una. Muchas gracias
